prog/dftb+/lib_dftbplus/main.F90

```diff
@@ -6247,6 +6247,80 @@ subroutine getReksInitialSettings(env, denseDesc, h0, over, neighbourList, &
   end subroutine getReksInitialSettings
 
 
+  !> Normalize eigenvectors with unitary transformation
+  subroutine renormalizeEigenvecs(env, electronicSolver, eigvecsReal, reks)
+
+    use dftbp_blasroutines, only : gemm
+
+    !> Environment settings
+    type(TEnvironment), intent(inout) :: env
+
+    !> Electronic solver information
+    type(TElectronicSolver), intent(inout) :: electronicSolver
+
+    !> Eigenvectors on eixt
+    real(dp), intent(inout) :: eigvecsReal(:,:,:)
+
+    !> data type for REKS
+    type(TReksCalc), intent(inout) :: reks
+
+    real(dp), allocatable :: tmpMat(:,:)
+    real(dp), allocatable :: tmpS(:,:)
+    real(dp), allocatable :: tmpC(:,:)
+    real(dp), allocatable :: tmpEigen(:)
+    real(dp), allocatable :: unitaryMat(:,:)
+
+    integer :: nOrb, ii
+
+    nOrb = size(reks%overSqr,dim=1)
+
+    allocate(tmpMat(nOrb,nOrb))
+    allocate(tmpS(nOrb,nOrb))
+    allocate(tmpC(nOrb,nOrb))
+    allocate(tmpEigen(nOrb))
+    allocate(unitaryMat(nOrb,nOrb))
+
+    ! Calculate CSC = C^T_old * S_new * C_old
+    tmpMat(:,:) = 0.0_dp
+    call gemm(tmpMat, eigvecsReal(:,:,1), reks%overSqr, transA='T')
+    tmpC(:,:) = 0.0_dp
+    call gemm(tmpC, tmpMat, eigvecsReal(:,:,1))
+
+    ! Make an identity matrix
+    tmpS(:,:) = 0.0_dp
+    do ii = 1, nOrb
+      tmpS(ii,ii) = 1.0_dp
+    end do
+
+    ! Diagonalize CSC to obtain a unitary matrix, U = CSC^(-1/2)
+
+    tmpEigen(:) = 0.0_dp
+    call env%globalTimer%startTimer(globalTimers%diagonalization)
+    ! tmpC becomes eigenvectors (X) of CSC
+    call diagDenseMtx(electronicSolver, 'V', tmpC, tmpS, tmpEigen)
+    call env%globalTimer%stopTimer(globalTimers%diagonalization)
+
+    ! Make inverse square root matrix consisting of eigenvalues (s) of CSC
+    tmpS(:,:) = 0.0_dp
+    do ii = 1, nOrb
+      tmpS(ii,ii) = 1.0_dp / sqrt(tmpEigen(ii))
```

Comment:
What happens in the case of of eigenvalues being ~0 ? Presumably if S_new is well conditioned, this should not be a problem, but it might be worth trapping for this.

Reply:
Then, how about to print warning message (or error) when the eigenvalues are close to zero? In test case (ethylene molecule), the eigenvalues are almost ~1.0 at many geometry steps and its maximum (minimum) values are ~1.5 (~0.6) at earlier step. Probably, 1.9 (0.1) would be proper standard? If you agree this, then I will add this check line.

Reply:
From additional test (commit 88af4af), the eigenvalues of ~0.3 or ~3.9 give correct convergence, so it is ambiguous to find a proper standard for eigenvalues in this case.

Reply:
Numerically it is only if the magnitude is below epsilon that there could be a problem. This is possible for badly behaved overlap, but probably just testing for this and stopping or just mildly regularize 1.0_dp/(max(sqrt(tmpEigen(ii)),epsilon(0.0_dp))